src/hotspot/share/runtime/arguments.cpp

```diff
@@ -744,20 +745,83 @@ bool Arguments::verify_special_jvm_flags(bool check_globals) {
 }
 #endif
 
-// Parses a size specification string.
-bool Arguments::atojulong(const char *s, julong* result) {
-  julong n = 0;
+template <typename T, ENABLE_IF(std::is_signed<T>::value), ENABLE_IF(sizeof(T) == 4)> // signed 32-bit
+static bool parse_integer_impl(const char *s, char **endptr, int base, T* result) {
+  STATIC_ASSERT(sizeof(long) >= 4); // need to have enough bits
+  errno = 0; // errno is thread safe
+  long v = strtol(s, endptr, base);
+  if (errno != 0 LP64_ONLY(|| v < min_jint || v > max_jint)) {
+    // long is 64-bit on LP64, so we need explicit range check.
+    return false;
```

Comment:
Windows is actually LLP64 rather than LP64 (but we generally pretend otherwise), so that `long` is 32bits on Windows.  That's why we avoid the use of `long` in shared code.  That makes the range tests tautological, and potentially warned about.  (gcc is known to warn about some tautological tests - I don't know if Visual Studio does.)  Maybe instead use `strtoll` and an unconditional range check?